presto-main/src/main/java/com/facebook/presto/dispatcher/QueuedStatementResource.java

```diff
@@ -0,0 +1,436 @@
+    }
+
+    @GET
+    @Path("/v1/statement/queued/{queryId}/{token}")
+    @Produces(APPLICATION_JSON)
+    public void getStatus(
+            @PathParam("queryId") QueryId queryId,
+            @PathParam("token") long token,
+            @QueryParam("slug") String slug,
+            @QueryParam("maxWait") Duration maxWait,
+            @HeaderParam(X_FORWARDED_PROTO) String xForwardedProto,
+            @Context UriInfo uriInfo,
+            @Suspended AsyncResponse asyncResponse)
+    {
+        Query query = getQuery(queryId, slug);
+
+        // wait for query to be dispatched, up to the wait timeout
+        ListenableFuture<?> futureStateChange = addTimeout(
+                query.waitForDispatched(),
+                () -> null,
+                WAIT_ORDERING.min(MAX_WAIT_TIME, maxWait),
+                timeoutExecutor);
+
+        // when state changes, fetch the next result
+        ListenableFuture<QueryResults> queryResultsFuture = Futures.transform(
+                futureStateChange,
+                ignored -> query.getQueryResults(token, uriInfo, xForwardedProto),
+                responseExecutor);
+
+        // transform to Response
+        ListenableFuture<Response> response = Futures.transform(
+                queryResultsFuture,
+                queryResults -> Response.ok(queryResults).build(),
+                directExecutor());
+        bindAsyncResponse(asyncResponse, response, responseExecutor);
+    }
+
+    @DELETE
+    @Path("/v1/statement/queued/{queryId}/{token}")
+    @Produces(APPLICATION_JSON)
+    public Response cancelQuery(
+            @PathParam("queryId") QueryId queryId,
+            @PathParam("token") long token,
+            @QueryParam("slug") String slug)
+    {
+        getQuery(queryId, slug)
+                .cancel();
+        return Response.noContent().build();
+    }
+
+    private Query getQuery(QueryId queryId, String slug)
+    {
+        Query query = queries.get(queryId);
+        if (query == null || !query.getSlug().equals(slug)) {
+            throw badRequest(NOT_FOUND, "Query not found");
+        }
+        return query;
+    }
+
+    private static URI getQueryHtmlUri(QueryId queryId, UriInfo uriInfo, String xForwardedProto)
+    {
+        return uriInfo.getRequestUriBuilder()
+                .scheme(getScheme(xForwardedProto, uriInfo))
+                .replacePath("ui/query.html")
+                .replaceQuery(queryId.toString())
+                .build();
+    }
+
+    private static URI getQueuedUri(QueryId queryId, String slug, long token, UriInfo uriInfo, String xForwardedProto)
+    {
+        return uriInfo.getBaseUriBuilder()
+                .scheme(getScheme(xForwardedProto, uriInfo))
+                .replacePath("/v1/statement/queued/")
+                .path(queryId.toString())
+                .path(String.valueOf(token))
+                .replaceQuery("")
+                .queryParam("slug", slug)
```

Comment:
Just double check: now slug is provided as a parameter, while in the old PR, slug is part of the path? (See https://github.com/prestodb/presto/pull/12801/commits/b84661660cb76f9247f1a9b21a85a1b9062f8ce2#diff-ef10213de38c18dc49c5303d7eea77a4R234)

Reply:
Yes, I saw from our commit history we did this because our HTTP logging seems to exclude query params from the logs.